Consensus Reactor communication specification (#74)
cometbft/cometbft: spec/consensus/reactor/README.md

# Consensus Reactor

> CometBFT is a state machine replication framework and part of the stack used in the Cosmos ecosystem to build distributed applications, such as the Cosmos SDK.

Reviewer: I wouldn't define the Cosmos SDK as an app. Consider mentioning the Cosmos Hub instead

CometBFT uses its northbound API, ABCI, to communicate with applications.
South of CometBFT is the OS' network stack.

CometBFT is implemented in a modular way, separating protocol implementations into **reactors**.
Reactors communicate with their counterparts on other nodes using the P2P layer, through what we will call the **P2P-I**.
This set of documents focuses on the interactions between the P2P layer and the Consensus Reactor.

```
                                                      SDK Apps
                                                   ==================
 Applications                                        Cosmos SDK    
======================================ABCI============================ ┐
 [Mempool Reactor] [Evidence Reactor] [Consensus Reactor] [PEX] [...]  |

> - - - - - - - - - - - - - - - - P2P-I - - - - - - - - - - - -- - - - - | Tendermint
>                                   P2P                                  | Core

Reviewer: ```suggestion
- - - - - - - - - - - - - - - - P2P-I - - - - - - - - - - - -- - - - - | CometBFT
                                  P2P                                  |
```

Exchanges between CONS and GOSSIP happens through the **GOSSIP-I**.

```
...
==========ABCI=========
                         ┐
  |      CONS       |    |
  |.....GOSSIP-I....|    |  Consensus Reactor
  |     GOSSIP      |    |
                         ┘
- - - - - P2P-I - - - -
         P2P
=======================
    Network Stack
```

The overall goal here is to specify the following:
1. Provide an english specification of
    * what CONS requires from and provides to GOSSIP (GOSSIP-I);
    * what GOSSIP requires from and provides to CONS (GOSSIP-I); and,
    * what GOSSIP requires from and provides to P2P (P2P-I) in order to satisfy CONS' needs.
2. Provide equivalent Quint specifications, used to mechanically check the properties
3. Provide an english description of how the current implementation matches (or not) the specified behavior, that is, a very loose refinement mapping.

# Outline

The specification is divided in multiple documents
* [reactor.md](./reactor.md): specification in English
* [reactor.qnt](./reactor.qnt): corresponding specifications in [Quint](https://github.com/informalsystems/quint)

> * [implementation.md](./implementation.md): a description of what is currently implemented in Tendermint Core, in English.

Reviewer: ```suggestion
* [implementation.md](./implementation.md): a description of what is currently implemented in CometBFT, in English.
```

* [implementation.qnt](./implementation.qnt): Quint model of current behavior, for model checking of provided properties.

# Conventions

> * MUST, SHOULD, MAY...

Reviewer: ```suggestion
* MUST, SHOULD, MAY... according to RFC2119.
```
